containers/hawkular-metrics-index-truncator/src/main/java/org/hawkular/openshift/truncator/Truncator.java

```diff
@@ -0,0 +1,144 @@
+        forceTruncation = Boolean.getBoolean("hawkular.metrics.force-truncation");
+        keyspace = System.getProperty("hawkular.metrics.cassandra.keyspace", "hawkular_metrics");
+        namespace = System.getProperty("hawkular.metrics.namespace", "openshift-infra");
+        cassandraConnectionMaxDelay = Long.getLong("hawkular.metrics.cassandra.connection.max-delay", 30);
+        cassandraConnectionMaxRetries = Integer.getInteger("hawkular.metrics.cassandra.connection.max-retries", 5);
+        truncatorMaxRetries = Integer.getInteger("hawkular.metrics.truncator.max-retries", 5);
+        truncatorMaxDelay = Long.getLong("hawkular.metrics.truncator.max-retries", 5*1000);
+    }
+
+    private String getToken() throws IOException {
+        return new String(Files.readAllBytes(Paths.get(TOKEN_FILE_LOCATION)));
+    }
+
+    public void truncate() throws IOException, InterruptedException, RuntimeException {
+
+        String token = this.getToken();
+        openshiftClient = new OpenshiftClientWrapper(this.namespace, this.KUBERNETES_MASTER_URL, token);
+
+        Pod[] cassandraPods = openshiftClient.getPodsWithPrefix(this.cassandraPodPrefix);
+        ArrayList<String> cassandraHosts = new ArrayList<>();
+        Arrays.stream(cassandraPods).forEach((pod) -> {
+            cassandraHosts.add(pod.getIP());
+        });
+
+        if (cassandraPods.length > 0 || this.forceTruncation) {
+
+            if (openshiftClient.hasLargePartition(this.partitionThreshold)) {
+                IndexTruncator truncator = new IndexTruncator(cassandraHosts, this.keyspace, this.useSSL);
+                truncator.setCassandraConnectionMaxDelay(cassandraConnectionMaxDelay);
+                truncator.setCassandraConnectionMaxRetries(cassandraConnectionMaxRetries);
+                // Get current replicas
+                int metricsReplicas = openshiftClient.getReplicationControllerScale(this.metricsRC);
+                int heapsterReplicas = openshiftClient.getReplicationControllerScale(this.heapsterRC);
+
+                // Scale to zero
+                openshiftClient.scaleReplicaitonController(this.heapsterRC, 0);
+                openshiftClient.scaleReplicaitonController(this.metricsRC, 0);
+
+                // Truncate
+                truncator.execute();
+
+                // Restore RC, this should rebuild indices.
+                openshiftClient.scaleReplicaitonController(this.heapsterRC, heapsterReplicas);
+                openshiftClient.scaleReplicaitonController(this.metricsRC, metricsReplicas);
+            } else {
+                log.info("Truncation was not necessary, skipping.");
+            }
+        } else {
+            throw new RuntimeException("No cassandra pods available, Is Hawkular Metrics running?");
+        }
+
+    }
+
+    public void run() {
+        int retryCounter = 0;
+        while (retryCounter < truncatorMaxRetries) {
+            try {
+                this.truncate();
```

Comment:
this cycle needs a `break` or something, otherwise it will run forever even if truncation succeeds (yeah, forever, because `retryCounter` will not increment)

Reply:
I put a return, also address the restoration of replicas on failure.

Reply:
I think this is incorrect now, imagine that `truncate()` is called, it detects the current number of metrics/heapster replicas, scales them both to 0 and then `truncator.execute()` fails, so when we retry, `truncate()` will again check the number of replicas but this will be zero and we will overwrite the original (correct) numbers with zeros, therefore the scale back will not work. The number of replicas (`metricsReplicas` and `heapsterReplicas`) should be checked only once during the first call of `truncate()`.

Reply:
I added a validation, if the scale returns zero I don't set `metricsReplicas` or `heapsterReplicas`